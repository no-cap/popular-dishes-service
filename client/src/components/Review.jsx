import React from 'react';
import StarBox from './starbox.jsx';
import Star from './star.jsx';
import PersonIcon from './personIcon.jsx';
import StarType from '../components-style/TypesOfStars.jsx';

import { 
    SingleReview, 
    PhotoCropper, 
    ProfilePic, 
    UserName, 
    TinyPairIcon, 
    FriendCount, 
    TinyStar, 
    UserReviews,
    StarContainer,
    Date,
    ReviewText
} from '../components-style/SingleReview.jsx'

class Review extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            text: this.props.data.text,
            readMore: false
        };
        this.onClickHandler = this.onClickHandler.bind(this);
        this.dateFormatter = this.dateFormatter.bind(this);
    }
    // handles clicking on Read more / Read less
    onClickHandler(e) {
        this.setState({ readMore: !this.state.readMore })
    }
    // Formats Dates
    dateFormatter(date) {
        var arr = date.slice(0, 11).split("-");
        if (arr[1][0] === '0') {
            var dateWithZero = arr[1] + '/' + arr[0] + '/' + arr[2];
            return dateWithZero.slice(1);
        } else {
            return arr[1] + '/' + arr[0] + '/' + arr[2];
        }
    }

    render() {
        return (
            <SingleReview >
                <PhotoCropper> <ProfilePic src={this.props.data.userphoto} /> </PhotoCropper>
                <UserName> {this.props.data.username} </UserName>
                <TinyPairIcon > 
                    <PersonIcon /> 
                </TinyPairIcon>
                <FriendCount > {this.props.data.friends} </FriendCount>
                <TinyStar> 
                    <StarType type="mini" rating={4}> <Star/> 
                </StarType> </TinyStar>
                <UserReviews> {this.props.data.reviews} </UserReviews>
                <StarContainer> 
                    <StarBox rating={this.props.data.rating} /> 
                </StarContainer>
                <Date> {this.dateFormatter(this.props.data.date)} </Date>
                <ReviewText>
                    <div id={this.props.data.user}>{(this.state.readMore) ? this.state.text : this.state.text.split(" ").slice(0, 22).join(" ")}</div>
                    {(this.props.data.text.split(" ").length > 22) ? <button onClick={this.onClickHandler}> {(this.state.readMore) ? 'Read less' : 'Read more'} </button> : null}
                </ReviewText>
            </SingleReview>
        )
    }
}

export default Review;


// { JSON.stringify(this.props.data) }