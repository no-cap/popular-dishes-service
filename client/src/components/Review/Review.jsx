import React from 'react';
import StarBox from '../Starbox/Starbox';
import Star from '../Starbox/Star';
import PersonIcon from '../PersonIcon/PersonIcon';
import StarType from '../Starbox/StarStyles';

import {
  SingleReview,
  PhotoCropper,
  ProfilePic,
  UserName,
  TinyPairIcon,
  FriendCount,
  TinyStar,
  UserReviews,
  StarContainer,
  Date,
  ReviewText,
} from './ReviewStyles';

class Review extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      text: this.props.data.text,
      readMore: false,
    };
    this.onClickHandler = this.onClickHandler.bind(this);
    this.dateFormatter = this.dateFormatter.bind(this);
  }

  // handles clicking on Read more / Read less
  onClickHandler(e) {
    this.setState({ readMore: !this.state.readMore });
  }

  // Formats Dates
  // eslint-disable-next-line class-methods-use-this
  dateFormatter(date) {
    const arr = date.slice(0, 11).split('-');
    if (arr[1][0] === '0') {
      const dateWithZero = `${arr[1]}/${arr[0]}/${arr[2]}`;
      return dateWithZero.slice(1);
    }
    return `${arr[1]}/${arr[0]}/${arr[2]}`;
  }

  render() {
    return (
      <SingleReview>
        <PhotoCropper>
          <ProfilePic src={this.props.data.userphoto} />
        </PhotoCropper>
        <UserName>{this.props.data.username}</UserName>
        <TinyPairIcon>
          <PersonIcon />
        </TinyPairIcon>
        <FriendCount>{this.props.data.friends}</FriendCount>
        <TinyStar>
          <StarType type="mini" rating={4}><Star /></StarType>
        </TinyStar>
        <UserReviews>{this.props.data.reviews}</UserReviews>
        <StarContainer>
          <StarBox rating={this.props.data.rating} />
        </StarContainer>
        <Date>{this.dateFormatter(this.props.data.date)}</Date>
        <ReviewText>
          <div id={this.props.data.user}>{(this.state.readMore) ? this.state.text : this.state.text.split(' ').slice(0, 22).join(' ')}</div>
          {(this.props.data.text.split(' ').length > 22) ? (
            <button onClick={this.onClickHandler}>{(this.state.readMore) ? 'Read less' : 'Read more'}</button>
          ) : null}
        </ReviewText>
      </SingleReview>
    );
  }
}

export default Review;
