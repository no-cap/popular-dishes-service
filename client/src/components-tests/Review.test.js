import React from 'react';
import { shallow, mount } from 'enzyme';
import Review from '../components/Review.jsx';
// data for one user
var singleData =
{
  user: 'Bob',
  text: 'I love corn cheese'
}

// all the tests
describe('Review Unit test', () => {
  // just seeing if the component even renders correctly
  describe('Review should render correctly', () => {
    test('Review should render correctly in "debug" mode', () => {

      const wrapper = shallow(<Review data={singleData} />);

      expect(wrapper).toMatchSnapshot();
    });

    test('Review should have only 1 div', () => {

      const wrapper = mount(<Review data={singleData} />);

      expect(wrapper.find('button')).toHaveLength(1);
    });
  })
  // making sure the component's click handler reacts correctly
  describe('Review Unit Interaction test', () => {
    describe('onClickHandler', () => {
      test('it should invoke the correct method when a change event is emitted', () => {
        const mockClickHandler = jest.fn();
        const wrapper = shallow(<Review data={singleData}/>);
        wrapper.instance().onClickHandler = mockClickHandler;
        wrapper.instance().forceUpdate();
        wrapper.find('button').simulate('click');
        expect(mockClickHandler).toHaveBeenCalled();
      });
      // testing that click button goes from true to false and vice versa
      test('it should correctly update the state when a change event is emitted', () => {
        const wrapper = shallow(<Review data={singleData} />);
        wrapper.find('button').simulate('click');
        expect(wrapper.instance().state.readMore).toBe(true);
        wrapper.find('button').simulate('click');
        expect(wrapper.instance().state.readMore).toBe(false);
      });
    });
  });

})
