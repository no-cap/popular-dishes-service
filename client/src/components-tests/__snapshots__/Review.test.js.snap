// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Review Unit test Review should render correctly Review should render correctly in "debug" mode 1`] = `ShallowWrapper {}`;
